bert/src/tokenizer/huggingface.rs

```diff
@@ -0,0 +1,198 @@
+    use_type_ids: bool,
+}
+
+impl Tokenize for Tokenizer {
+    fn new<P>(config: &Config<Self, P>) -> Result<Self, Error> {
+        let mut tokenizer = HfTokenizer::from_file(config.dir.join("tokenizer.json"))?;
+        let padding_token = config.extract::<String>("tokenizer.tokens.padding")?;
+        let padding = PaddingParams {
+            strategy: PaddingStrategy::Fixed(config.token_size),
+            direction: PaddingDirection::Right,
+            pad_to_multiple_of: None,
+            pad_id: tokenizer
+                .token_to_id(&padding_token)
+                .ok_or("missing padding token")?,
+            pad_type_id: 0,
+            pad_token: padding_token,
+        };
+        let truncation = TruncationParams {
+            direction: TruncationDirection::Right,
+            max_length: config.token_size,
+            strategy: TruncationStrategy::LongestFirst,
+            stride: 0,
+        };
+        tokenizer.with_padding(Some(padding));
+        tokenizer.with_truncation(Some(truncation));
+        let use_type_ids = config.extract::<Dict>("model.input")?.len() > 2;
+        let add_special_tokens = config.extract::<bool>("tokenizer.add-special-tokens")?;
+        Ok(Tokenizer {
+            hf_tokenizer: tokenizer,
+            add_special_tokens,
+            use_type_ids,
+        })
+    }
+
+    fn encode(&self, sequence: impl AsRef<str>) -> Result<Encoding, Error> {
+        let tokens = self
+            .hf_tokenizer
+            .encode(sequence.as_ref(), self.add_special_tokens)?;
+        let token_ids: Vec<u32> = tokens.get_ids().to_vec();
+        let attention_mask = tokens.get_attention_mask().to_vec();
+        let type_ids = tokens.get_type_ids().to_vec();
+        let array_from =
+            |slice: &[u32]| Array2::from_shape_fn((1, slice.len()), |(_, i)| i64::from(slice[i]));
+
+        Ok(Encoding {
+            token_ids: array_from(&token_ids),
+            attention_mask: array_from(&attention_mask),
+            type_ids: if self.use_type_ids {
+                Some(array_from(&type_ids))
+            } else {
+                None
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use xayn_test_utils::asset::{e5_mocked, smbert_mocked};
+
+    use super::*;
+
+    #[test]
+    fn test_smbert() {
+        let config = Config::new(smbert_mocked().unwrap()).unwrap();
+        let tokenizer = Tokenizer::new(&config).unwrap();
+        let encoding = tokenizer
+            .encode("These are normal, common EMBEDDINGS.")
+            .unwrap();
+        assert!(encoding.token_ids.shape() == [1, 257]);
+        let first_ids = encoding
+            .token_ids
+            .iter()
+            .take(20)
+            .copied()
+            .collect::<Vec<_>>();
+        assert!(
+            first_ids
+                == vec![
+                    2, 4538, 2128, 8561, 1, 6541, 69469, 2762, 5, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+                ]
+        );
```

Comment:
:art: iterating and collecting into a vector is definitely a possibility to compare, we can improve readability for example via slicing though.
```suggestion
        assert_eq!(
            encoding.token_ids.slice(s![0, ..20]),
            arr1(&[2, 4538, 2128, 8561, 1, 6541, 69469, 2762, 5, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
        );
```
the slicing returns an array, that's why we need to wrap the expected values in one instead of a vec. the items can be imported via `use ndarray::{arr1, s};`. similarly for the other tests.